---
title: "DECO report from heterogeneity analysis"
author:
- Francisco Jose Campos-Laborie
- Jose Manuel Sanchez-Santos
- and Javier De Las Rivas
- Cancer Research Centre. Bioinformatics and Functional Genomics Lab.
- University of Salamanca, Salamanca. Spain
header-includes:
    - \usepackage{fancyhdr}
    - \pagestyle{fancy}
    - \fancyhead{}
    - \fancyhead[RO,LE]{\textbf{DECO R package}}
    - \fancyhead[LO,RE]{\textsl{\leftmark}}
geometry: margin=1in
output:
  pdf_document:
    dev: pdf
    fig_caption: yes
    highlight: tango
    toc: yes
  html_document:
    dev: png
    fig_caption: yes
    fig_height: 6
    fig_width: 7
    theme: united
    toc: yes
  word_document:
    dev: png
subtitle: DEcomposing heterogeneous Cohorts using Omic profiling data.
tags:
- nothing
- nothingness
abstract: DECO carries out a comprehensive analysis of how both features and samples are structured along the dataset, revealing some hidden dependence between them. This vignette tries to illustrate some information and plots generated by *decoReport* R function. 
---

<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{deco}
\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{pdfpages}
\usepackage[a4paper,margin=1in]{geometry}
\usepackage{fancyhdr}
\pagestyle{fancy}
\lhead{Lecture 1}
\usepackage[justification=centering,margin=1cm]{caption}
\captionsetup{
  format=plain,
  margin=0em,
  labelsep=newline,
  justification=raggedright,
  singlelinecheck=false
}
%\usepackage{subcaption}
-->
<style type="text/css">
table{ 
    float:center;
    display: block;
    margin-left: auto;
    margin-right: auto;
}
img { 
    max-width: 90% !important; 
    float:center;
    display: block;
    margin-left: auto;
    margin-right: auto;
}
table img{ 
    max-width: 300px !important; 
   }
.subtitle{
    font-size: 95%; 
    margin-left: 3in;
    margin-right: 3in;
}
</style>

```{r pre,echo=FALSE,results='hide'}
library(knitr)
opts_chunk$set(warning=FALSE,message=FALSE,cache=TRUE,fig.path='figures/',tidy=FALSE,tidy.opts=list(width.cutoff=60))
```

---

\pagebreak

# Analysis report generated by *decoReport()*

An extended PDF report can be generated as summary of the analysis with main relevant information. The `decoReport()` function uses *deco* object generated by previous function to create this report.

```{r, warning=FALSE}
### microarray results
decoReport(deco_results, sub_binary, pdf.file = "Report.pdf",
           info.sample = pData(ALCLdata)[,c("Type","Blood.paper"), drop = FALSE],
           cex.names = 0.3, print.annot = TRUE)
```

Before showing how the report is organized, several of the input parameters should be mentioned. First, phenotypic information of samples or feature's information (i.e. gene functional annotation) can be included in `info.sample` and `info.gene` as a `data.frame` object with samples or features IDs as rownames. To find out associations of this information to any subclass, it would be plotted in all heatmaps.

The differential features found by DECO were characterized with various statistics (summarized in`deco@featureTable` slot) that could be used to highlight features with significant omic profiles. Two tables with different rankings for *complete*/*majority* changes and *minority*/*mixed* changes will be shown in this PDF.

# 1. Summary of DECO

First page of the report is a summary of the analysis. It is composed by general information about analysis settings and main results. 

\begin{figure}[h]
\centering
\frame{\includegraphics{figures/overview.pdf}}
\caption{Summary page of PDF report generated by DECO algorithm.}
\end{figure}

Then, at top-left place, report shows relevant information about global analysis like what type of analysis was run (*binary*, *unsupervised*, *multiclass*), how many combinations have at least one differential features (*positive iterations*) and how many have been included in NSCA after double-threshold of repeats: *pos.rep* threshold for the minimum number of *Repeats*. A good approach to how much distinct are two initial classes compared (if *binary* analysis) remains in the percentage of *positive iterations* more than how many differential features have been found, because these features could be only associated to a specific subclass.

Moreover, the percentage of variability explained by NSCA and p-value associated is shown. If **p-value** is no significant (**p-value > 0.05**), we can not assume that the assymetric association between genes and samples is good. In this way, some reasons may be:

- The number of `iterations` run in **RDA** function `decoRDA()` is not enough to explain heterogeneity. It would be necessary to increase this parameter to **reanalyze the dataset and ruling out other reasons**.

- There are not **any assymetric association** between features and samples, so subclasses found by `decoNSCA()` should not be take into consideration (i.e. comparisons among homogeneous classes of samples). 

\begin{figure}[h]
\centering
\includegraphics[width=3in,height=3in]{figures/information1.jpg}
\caption{RDA and NSCA main information.}
\end{figure}

At middle of left side (Fig. 3), a short table with top 10 features is disposed. If the feature annotation with official gene symbol and direction of differential expression (if *binary* analysis) are available, they would also appear in this table.

\begin{figure}[h]
\centering
\includegraphics[width=3in,height=3in]{figures/information2.jpg}
\caption{Top-10 features selected by the specified way of rank.}
\end{figure}

At bottom of this report's page (Fig. 4), the report shows a summary of subclasses found with up-regulated features within each subclass. Control subclasses will be colored **blue scale**, while case subclasses will be colored **red-orange scale**. However, if NSCA p-value is no significant for any class or if any feature is related to, corresponding subclasses may be less significant and they would be colored in grey. 

\begin{figure}[h]
\centering
\includegraphics[width=4in,height=3.5in]{figures/information3.jpg}
\caption{Short summary subclasses found. The number of samples and features included are shown and if it is statistical relevant or not.}
\end{figure}

At right panel of this summary page (Fig. 5), the sample membership to subclasses is shown. All samples included in NSCA will be related to one subclass but if any differential feature was found for some samples, they will labelled as *NoSubclass*. As we can see, all samples included in the analysis are labelled with a number that will remain stable along the report. To identify a particular sample, this labels will reference them.

Moreover, we provided in a separated page the color code (Fig. 6) used along the report to refer each phenotype and subclass label.

\begin{figure}[h]
\centering
\includegraphics[width=6in,height=3.5in]{figures/information5.jpg}
\caption{The color code corresponding to each sample label containing relevant information.}
\end{figure} 

\pagebreak

# 2. Differential event distribution

We provide a graphic measure of how many differential events are summarized by each sample. Due to the properties of the differential analysis, this information would be relevant as a indirect way to determine how distant two group samples are. If we take a look at figure 7, we could see how green-labeled samples (corresponding to PTCL subclass of example data (PTCL samples of ALCL dataset) accumulated more differential repeats. It means that any iteration including samples from that subcategory shows more differential expression contrasting to positive ALK samples, so those samples are quite distant while no-PTCL ones are less.

\begin{figure}[p]
\centering
\frame{\includegraphics{figures/repeat_count.pdf}}
\caption{The differential events accumulated per feature and percentage of samples which have one differential event at least. Color-code corresponds to defined threshold after RDA step (red points are removed features).}
\end{figure}

# 3. Goodman and Kruskal's Tau sample contribution

Inherited from NSCA, each sample contribution to global differential signal can be assesed. In this way, the *Goodman and Kruskal's Tau* or $\tau$ statistic from [NSCA](#references) [7] can be decomposed as addends to determine individual contribution. If any sample accumulates an unusual amount of $\tau$ (or exclusive differential signal), it could be an outlier.

For this reason, in the report is included a boxplot representation (Fig. 8) of these sample contributions to global *Goodman and Kruskal's Tau*. In the figure below, we can observe how sample number (reference number) is unusally deviated from the equitate contribution suppossed for all samples.

\begin{figure}[h]
\centering
\frame{\includegraphics{figures/tau_samp.pdf}}
\caption{Boxplot of Goodman and Kruskal's Tau sample contribution. Outlier values for any sample (or group of samples) indicates this sample accumulates more specific signal than others.}
\end{figure}

Moreover, at right side of this pdf page we have included the statistical analysis of $\tau$, so we could consider the results from NSCA step if its *p-value* is under 0.05 or 0.1. In that case, NSCA is not able to assign properly differential features to samples so that could have two reasons:

- All samples, or samples from a category if *binary* analysis was set up, have an homogeneous distribution of its DE signal, so any relevant gene-sample relationship has been found.

- Number of combinations or `iterations` set up by user in the *RDA* function was too low and NSCA has not enough information to solve the heterogeneous dataset. We recommend to **restart all the analysis increasing the number of `iterations`**.


# 4. Top 50 features

Next pages of report include a maximum of top 50 features (Fig. 9) ranked by *type.ranking* input method with relevant gene information:

\begin{figure}[p]
\centering
\frame{\includegraphics{figures/ranking.pdf}}
\caption{Top-50 features' ranking. IDs provided by user to show theirs feature profiles would be highlighted in red (ALK here).}
\end{figure}

1. *Ranking* marks the feature position in global ranking. Remember that two ways of rank features (*type.ranking* may be *generic* or *specific*) can be chosen if BINARY analysis was carried out.

2. *ID* corresponds to original IDs from omic data matrix (*data* input).

3. *UpDw* with direction of differential signal in case of *binary* analysis. Remember that: 
    + **UP** features are UP-regulated in cases and DOWN-regulated in control.
    + **DOWN** features are DOWN-regulated in cases and UP-regulated in control.
    + **MIXED** features are UP-regulated in both classes.

4. *Profile* indicates which kind of feature profile explains better feature behavior if **binary** analysis was carried out (Fig. 2). Remember that:
    + **Complete** profile corresponds to a almost perfect change between categories.
    + **Majority** profile defines the feature as changing in the majority of samples within case class in comparison with control.
    + **Minority** profile defines the feature as changing in the minority of samples within case class in comparison with control.
    + **Mixed** profile is related to a similar change within both classes analysed for this feature.

5. *overlap.Ctrl.Case* measures how much signal is overlapping between both categories of samples analysed if **binary** analysis was carried out (0 means non-overlapping signal while 1 means all signal is overlapping).

6. *Standard.Chi.Square* is measuring the consistency of the differential signal of each feature (from RDA step). Further information could be found in the original publication.

7. *Repeats* counter corresponds with the number of positive iterations (with differential signal under the p-value threshold chosen for RDA). More *Repeats* does not indicate best **class** biomarker because differential events of each feature could be explained just by some samples.

8. *Repeats.index* indicates the percentage of samples affected by any differential event of this feature. 

9. *delta.signal* reflects the difference between omic signal mean by class (control and case) and it only appears when *binary* analysis is computed.

10. *sd.Control* and *sd.Case* measures standard deviation of raw omic signal. Low standard deviation should indicate best classes biomarkers and penalizes specific subclasses biomarkers.

11. *Dendrogram.group* refers to feature pattern membership. Due to biclustering analysis on **h-statistic** matrix, feature dendrogram could be also divided into feature subgroups following same behavior. If *binary* analysis was run, feature would have been assigned to one pattern in each category.

12. *h.Range* refers to the ability of separating subclasses (within or not a category of samples). It is the maximum difference between *h statistics* per subclass per feature.

\pagebreak

# 5. Differential events threshold per feature

Due to RDA provides huge amount of information, we need to apply a filter in order to remove all noisy signal from low expressed features. In this way, we have designed an easy way to remove features which are not strongly related to any sample or subclass of samples. As we explain in `decoNSCA()` R function, both `rep.thr` and `samp.perc` input variables indicates how this threshold has to be imposed. 

Inside the frequency matrix generated by RDA step, a noisy feature would show most of differential events equally distributed along samples showing lower values than a great biomarker which is supposed to mark all the samples with higher amounts of differential events (Fig. 10). Adittionally, a specific feature related just to a subclass of samples would amount most of its differential events on those samples, so it would also have higher amounts of differential events. 

Thus, we decomposed **Repeats or differential events** vector per feature along samples in: **(i)** minimum amount of differential events accumulated per sample or `rep.thr`; and **(ii)** minimum percentage of samples showing this `rep.thr` at least, called `samp.perc`.

\begin{figure}[H]
\centering
\frame{\includegraphics{figures/rep_thr.pdf}}
\caption{Plot comparing differential events or repeats per feature VS percentage of samples affected one time at least. Here, color code corresponds to current threshold imposed (rep.thr=3 and samp.perc=0.05), where red-color means filtered features. We can see how ALK gene has the greatest correlation.}
\end{figure}

# 6. Differential signal and profile assigment

Next, we represent `overlap.Ctrl.Case` VS `delta.signal` per feature if *binary* analysis was carried out and, here, the circle size corresponds to relative amount of `Standard.Chi.Square` accumulated per feature along all RDA iterations.

\begin{figure}[H]
\centering
\frame{\includegraphics{figures/od_plot.pdf}}
\caption{Plot comparing overlap and delta signal per feature. This plot is similar to **volcano plots** where p-value and fold-change are represented in a two-dimensional plot.}
\end{figure}

# 7. Subclasses found

In the next page, dendrogram(s) showing subclasses of samples (Fig. 12)  found based on **h-statistic** is shown. Moreover, Hubber's gamma coefficient used to cut dendrogram (if no number of subclasses was set up by user) is also indicated. Further information about how this dendrogram was cut is reported in the paper.

\begin{figure}[h]
\centering
\frame{\includegraphics{figures/dendr.pdf}}
\caption{Dendrogram(s) of subclasses found by DECO. Instead original names of samples, reference numbers are shown here.}
\end{figure}

# 8. Top-50 discriminant features based on **h statistic**

Because the **h-statistic** values growth for feature-sample dependence (positive value for up-regulated and negative for down-regulated), the subclass' assigment based on a feature presence could be derived. Provided that, range of values, mean and standard error of **h-statistic** per subclass of samples per feature are good estimators of how much discriminant is any feature. Thus, **h** range of values indicates maximum discriminant power of each statistic, while mean and ranking per subclass refer to specific subclass relationship.

Here, a table summarizing mean and ranking of top-50 features (Fig. 13) within a category of samples (top-50 based on **h.range**) is shown. Notice that any feature could be determined as discriminant for one or more subclasses found by DECO, so assigning feature membership just to one subclass may be inadequate.

\begin{figure}[h]
\centering
\frame{\includegraphics{figures/ranking_h.pdf}}
\caption{Table showing h statistic average per subclass per feature (gene). We can see how any gene is able to differentiate one or more subclasses (it would be high ranked within a subclass).}
\end{figure}

# 9. Biclustering analysis on **h-statistic**

## Heatmap of **h-statistic**

As we can see along all this vignette, **h-statistic** matrix obtained from NSCA and RDA is the main result to figure out all feature-sample dependence.

Unlike classical dimensionality reduction techniques, such as Principal Component Analysis (PCA) or Correspondence Analysis (CA), to disclose a means of displaying or summarising a dataset in two-dimensional graphical form, we aim solving heterogeneous data trying to compile **most of the variability** in a single context and not reducing it in a simple picture (biplot or tridimensional plot of principal axes). For this reason, the best way to show relations among feature and samples is using a biclustering technique (heatmap). 

Once we calculate our **h-statistic** matrix from omic data and NSCA step, we grouped samples with similar profiles (using all differential features after differential event threshold) using a hierarchical clustering, so the dendrogram obtained will be fixed to order samples properly in their corresponding subclasses. We can see (Fig.14) an example of heatmap and subclasses corresponding to ALCL dataset:

\begin{figure}[h]
\textbf{Biclustering heatmap on h statistic using features found by DECO}\par\medskip
\centering
\frame{\includegraphics{figures/heatmap1.pdf}}
\caption{Heatmap representation of h statistic values from DECO within the PDF report generated by decoReport() function.}
\end{figure}

#### Color scale of **h-statistic**

Due to our previously explained **h-statistic** summarizes both data dispersion within a feature and standarized inner product, its scale is platform-dependendent, so we provided its color-code and density plot from all or a category of samples (if **binary** analysis was carried out).

#### Sample info by user

To improve the visualization of coincidences among phenotypical information with subclasses found, the user can provide more information about samples (i.e. phenotypical information) or features (i.e. derived from functional enrichment analysis) introducing a `data.frame` in `info.sample` and `info.feature` inputs. It will be plot as color-bars under the dendrograms of both samples and features. 

In our example, we include some information about subtypes of ALK- ALCL's samples. The PCTL subtype (green coloured label) within these samples is quite distant (subclass 1) from others, as well as related literature describes it (Fig. 14).

#### Feature patterns

The feature's names printed along the heatmap could be original IDs from *data* input matrix or annotated names, if **print.annot** option is TRUE and we previously run annotation process in `decoRDA()` RDA function. In orange, we highlight IDs selected by user with **ids** input in `decoReport()` function. 

Additionally, feature patterns assignment (Fig. 14 at left) is printed and numbered along feature's dendrogram to facilitate the localization of any particular feature within heatmap. It is important to mention that `Dendrogram.order` positions are placed from top to bottom within feature's dendrogram, so first (number 1) position would be at top and last position at the bottom.

#### Differential feature direction and feature pattern membership

Due to **h-statistic** matrix is represented here, we have added labels with the **differential signal of each feature in these samples** if **binary** analysis. If a feature is green-labelled means that it is DOWN regulated for these samples, meanwhile a red-label means that it is UP regulated.

##### Sample number references

Instead to printing whole sample names below the heatmap, we use the number reference previously mentioned in the [summary page](## 1.1 Summary of DECO) of the report.

#### Binary and Unsupervised analysis

As mentioned above, each type of analysis design carries out different data treatmen within NSCA step (`decoNSCA()` function). If *binary* analysis was run, the `incidenceMatrix` input to the NSCA was split depending on sample labels (control and cases); meanwhile there was no splitting procedure with *unsupervised* or *multiclass* analysis, so all samples were included together. In this way, the *binary* analysis will generate a heatmap per class of samples while *unsupervised* will not.

## Heatmap of raw data matrix

As a proper way to contrast our results with original data, we proposed to compare final **h statistic** heatmap with an unsupervised heatmap of raw omic data (omic data here, Fig. 15). Thus, using the same features that previous heatmap does, we group samples and features in a similar way using original omic data:

\begin{figure}[h]
\textbf{Biclustering heatmap on RAW omic signal using features found by DECO}\par\medskip
\centering
\frame{\includegraphics{figures/heatmap2.pdf}}
\caption{Heatmap representation including all samples (for both binary and unsupervised analysis) and the same differential features that inner product heatmap does.}
\end{figure}

Here, we have used the most established color scale to represent gene expression: green-red scale. Additionally in all *binary* DECO analyses, it is represented by default both classes compared in **DECO.Contrast.design** labels (here, *pos* as control and *neg* as cases). As we can see, any subclass found by DECO are shown because here the *samples dendrogram* is depending on raw gene expression data.

# 10. Feature profile

As we explained in the Materials and Methods section of the publication, the omic profile representation of each differential feature is inherited from the **h-statistic** information. Once we calculated this vector per feature, we can group samples within any particular profile depending on its *h* values. Thus, each feature will have the same relevance for the characterization of those samples grouped together. 

If a **binary analysis** was carried out, this page is composed by several information about a particular feature: (**top-left**) a omic data profile with samples points increasingly ordered by **h-statistic** from left to right within a category (control or case), then, sample labels and subclass assigment are placed under this profile; (**top-right**) a representation of overlapping signal between both categories compared; (**bottom-left**) a reminder section of relevant information about color-code and statistics for this feature; (**bottom-right**) a barplot representation of **h-statistics** per subclass found, showing which subclasses are higher discriminated by this feature (higher or lower values).

In the examples shown below (Fig.16-19), we can observe how some samples in the case (or 1's) class are **up-regulated** and grouped together for both ERBB4 and COL6A5 gene, making a good example of [specific differential feature](#7. report analysis). 

\begin{figure}[p]
\textbf{ALK gene: profile section}\par\medskip
\centering
\includegraphics{figures/exp_profile.pdf}
\caption{Looking for local patterns within a omic profile using NSCA derived information and h statistic. This profile corresponds to ALK gene expression profile, which was used by authors to separate patients, so it is consider as an ideal example of COMPLETE change among categories. We can see how OVERLAP measurement (top-right) classifies perfectly this feature as COMPLETE. Moreover, h statistics per subclass found are close to 0, so there is not any specific signal related to a subclass of samples.}
\end{figure}

\begin{figure}[p]
\textbf{IL26 gene: profile section}\par\medskip
\centering
\includegraphics{figures/exp_profile3.pdf}
\caption{Looking for local patterns within a omic profile using NSCA derived information and h statistic. This profile corresponds to IL26 gene expression profile, which was assigned by DECO as a MAJORITY change profile between categories. Here, we show an example of a False Positive profile determined by SAM method as positive and discarded profile by COPA method. IL26 was placed as second best feature profile distinguishing classes by SAM, due to this large difference of signal; meanwhile, COPA discards this gene as a 'outlier gene'. We can see how the OVERLAP measurement (top-right) classifies this feature as MAJORITY. Moreover, h statistics per subclass found are distant to 0 for the new subclass, so there is dependencen between IL26 and this subclass of samples.}
\end{figure}

\begin{figure}[p]
\textbf{COL6A5 gene: profile section}\par\medskip
\centering
\includegraphics{figures/exp_profile2.pdf}
\caption{Looking for local patterns within a omic profile using NSCA derived information and h statistic. This profile corresponds to COL6A5 gene expression profile, which was discovered by authors as one of two biomarkers related to the new subclass of ALCL patients (marked in yellow within 'paper' labels), so it is consider as an ideal example of MINORITY change among categories. We can see how OVERLAP measurement (top-right) classifies perfectly this feature as MINORITY. Moreover, h statistics per subclass found are distant to 0 for the new subclass, so there is dependencen between COL6A5 and this subclass of samples.}
\end{figure}

\begin{figure}[p]
\textbf{ERBB4 gene: profile section}\par\medskip
\centering
\includegraphics{figures/exp_profile4.pdf}
\caption{Looking for local patterns within a omic profile using NSCA derived information and h statistic. This profile corresponds to ERBB4 gene expression profile, which was discovered by authors as one of two biomarkers related to the new subclass of ALCL patients (marked in yellow within 'paper' labels), so it is consider as an ideal example of MINORITY change among categories. We can see how OVERLAP measurement (top-right) classifies perfectly this feature as MINORITY. Moreover, h statistics per subclass found are distant to 0 for the new subclass, so there is dependencen between ERBB4 and this subclass of samples.}
\end{figure}

Furthermore, a box image with phenotypic information corresponding to `info.sample` input is included below each feature profile, revealing us how sample information is related to singular feature patterns found. In this way, any of the best specific biomarkers could be assigned to any particular phenotypical subgroups.

\clearpage

# 11. Variability explained

As final part of the report, it is informative to know how many NSCA dimensions were required to reach minimum threshold imposed with *v* input in `decoNSCA()` function. Then, we represent the cumulative percentage of variability explained by dimensions, so higher number of dimensions needed resembles more complexity or heterogeneity within our dataset (Fig. 27).

\begin{figure}[h]
\centering
\frame{\includegraphics{figures/var.pdf}}
\caption{Cumulative variability explained by NSCA for both (control and cases) classes.}
\end{figure}

In the figure, **red dashed line** highlights which percentage of variability have been explained by NSCA. As well as in previous results, this plot will be split for both classes in case of *binary* analysis.


# 12. Biplot and 3D representation of NSCA coordinates

We included a biplot and three-dimensional plot of NSCA coordinates for samples included in our analysis at final pages of this report. This representation would be split in control and case categories if *binary* or *supervised* analysis was run.

\begin{figure}[h]
\centering
\includegraphics{figures/biplot.pdf}
\caption{Biplot representation of the two major dimensions from NSCA.}
\end{figure}

\pagebreak

